### PR TITLE
Variable labels with packed column

### LINKED ISSUE #142: Variable labels with packed column / label attribute on data frame
In the dm package, we want to attach labels to table objects (https://github.com/cynkra/dm/pull/1888). Another use case is packed columns -- SPSS might not have them, still it's conceivable that users might want to attach labels to their packed columns.

The `var_label()` method for data frames iterates over columns by default, and returns the column labels. Would you consider offering a mode of operation where `var_label()` sets or returns the data frame's label?

### PR DESCRIPTION
**Pull request by larmarange**
fix #142

### CODE REVIEW

File: R/var_label.R [hunk 1]
```diff
@@ -2,17 +2,26 @@
...
 #' @param value a character string or `NULL` to remove the label
-#'  For data frames, it could also be a named list or a character vector
-#'  of same length as the number of columns in `x`.
+#'  For data frames, with `var_labels()`, it could also be a named list or a
+#'  character vector of same length as the number of columns in `x`.
 #' @param unlist for data frames, return a named vector instead of a list
 #' @param null_action for data frames, by default `NULL` will be returned for
 #' columns with no variable label. Use `"fill"` to populate with the column name
 #' instead, or `"skip"` to remove such values from the returned list.
+#' @param recursive if `TRUE`, will apply `var_label()` on packed columns
```
Comment (krlmlr) [inline R/var_label.R:11]: The tidyverse uses `recurse` as argument name in similar cases.

File: R/var_label.R [hunk 2]
```diff
@@ -2,17 +2,26 @@
...
+#'   `get_variable_labels()` is identical to `var_label()`.
+#'
+#'   For data frames, if you are using `var_label()<-` and if `value` is a
+#'   named list, only elements whose name will match a column of the data frame
+#'   will be taken into account. If `value` is a character vector, labels should
+#'   be in the same order as the columns of the data.frame.
+#'
+#'   If you are using `label_attribute()<-` or `set_label_attribute()` on a data
+#'   frame, the label attribute will be attached to the data frame itself, not
+#'   on a column of the data frame.
```
Comment (krlmlr) [inline R/var_label.R:24]: Typo?

```suggestion
#'   to a column of the data frame.
```

File: R/var_label.R [hunk 3]
```diff
@@ -224,8 +249,40 @@ set_variable_labels <- function(.data, ..., .labels = NA, .strict = TRUE) {
...
+#' @export
+get_label_attribute <- function(x) {
+  label_attribute(x)
+}
+
+#' @rdname var_label
+#' @export
+set_label_attribute <- function(x, value) {
+  label_attribute(x) <- value
+  x
```
Comment (krlmlr) [inline R/var_label.R:290]: I would expect the `<-()` function to call the setter. Besides being shorter, this *might* avoid extra duplication of R objects due to the way how `<-()` works.

**Review by krlmlr — COMMENTED** [review]
Thanks for looking into it! What do you think about calling the new accessor `get_label()` ? The "attribute" part seems to be an implementation detail.

**Comment by larmarange** [comment]
> Thanks for looking into it! What do you think about calling the new accessor `get_label()` ? The "attribute" part seems to be an implementation detail.

`label()` is already used by Hmisc.

`get_label()` and `set_label()` are already used by `sjlabelled()`, cf. https://strengejacke.github.io/sjlabelled/reference/get_label.html, but works similarly as `var_label()`

I would like to avoid conflict or to induce confusion as several packages have implemented labels features but with different approaches.

File: R/var_label.R (hunk 1 above)
Comment (larmarange) [reply R/var_label.R]: changed

File: R/var_label.R (hunk 3 above)
Comment (larmarange) [reply R/var_label.R:290]: I'm not sure to understand your point here. 

`set_label_attribute()`, as the other `set_*()` functions, is not supposed to changed the object, but rather to return the changed object.

**Comment by krlmlr** [comment]
My comment didn't seem to have gone through.

Have you seen https://adv-r.hadley.nz/functions.html#replacement-functions ?

You could write

```r
`label_attribute<-` <- set_label_attribute
```

with exactly the same semantics as your code.

**Comment by larmarange** [comment]
done through https://github.com/larmarange/labelled/commit/647b467a70ba910728906c35cb9a28371277e76c

**Pull request merged**